### PR TITLE
Move MPI and CUDA to Julia extensions

### LINKED ISSUE #4: Remove dependency on CUDA / KernelAbstractions in toplevel ClimaComms package
These deps can now live in sub packages that extend the toplevel abstract impl.

### LINKED ISSUE #74: Turn MPI and CUDA to extensions and standardize how to context and device and picked
As things are, everything that depends on ClimaComms inherits its dependencies: MPI and CUDA. These are not lightweight, especially CUDA. I think that extensions might work well here (This is what #4 proposed, and now that's a good way to implement that.
). 

The only downside is that downstream dependencies will have to explicitely depend and load MPI/CUDA. This would be quite painful to implement because we rely heavily on automatic selection of the context and the device.

The key functions are:
https://github.com/CliMA/ClimaComms.jl/blob/13f332d3770dd2591a79e2852933bc45f315f6ed/src/context.jl#L13-L32
and
https://github.com/CliMA/ClimaComms.jl/blob/13f332d3770dd2591a79e2852933bc45f315f6ed/src/devices.jl#L52-L74

These functions read some env variable, and if not available, they try to guess a resonable device. This is nice because for the most part we don't have to worry about setting anything and the code just runs (but also see #67, sometimes the guesses are incorrect). Unfortunately, I think that guessing the correct device/context is incompatible with using extensions because specific packages have to be loaded (eg `CUDA`). With the current implementation of PR #75, one would have to load `CUDA` to use a GPU on a GPU-capable device, otherwise the device would be set to CPU. This is particularly annoying in all CI runs, where we would have to add logic to handle the different cases and import the relevant packages.

Instead, I propose to deprecate the automatic selection of the device and let ClimaComms load the relevant module when asked. 
This might look like:
```julia
function device()
    env_var = get(ENV, "CLIMACOMMS_DEVICE", nothing)
    if !isnothing(env_var)
        if env_var == "CPU"
            return Threads.nthreads() > 1 ? CPUMultiThreaded() :
                   CPUSingleThreaded()
        elseif env_var == "CPUSingleThreaded"
            return CPUSingleThreaded()
        elseif env_var == "CPUMultiThreaded"
            return CPUMultiThreaded()
        elseif env_var == "CUDA"
           pkgid = Base.PkgId(Base.UUID("052768ef-5323-5732-b1bb-66c8b64840ba"), "CUDA")
           if !haskey(Base.loaded_modules, pkgid)
               try
                    Base.eval(Main, :(using CUDA))
                catch err
                     error("Cannot load CUDA.jl. Make sure it is included in your environment stack.")
                end
            end
            return CUDADevice()
        else
            error("Invalid CLIMACOMMS_DEVICE: $env_var")
        end
    end
end
```
This loads CUDA when `CLIMACOMMS_DEVICE = CUDA`. Now, the only responsability of the user is to ensure that they have CUDA in the environment they are using. The function will fail when the env variable `CLIMACOMMS_DEVICE` is not set. (This solution is currently loading CUDA into Main, but maybe it's best to load it within ClimaComms)

### PR DESCRIPTION
**Pull request by Sbozzolo**
Every package that depends on ClimaComms depends on CUDA and MPI. As a result, user of dowstream packages have to obtain CUDA and MPI even if they don't cannot or don't want to use either of the two.

This PR moves CUDA and MPI to extensions, opening the path to move support for CUDA and MPI in downstream packages to extensions too. 

The benefits of using extensions are reduced instantiation and load time (eg, in every GitHub Action runner), but also establishes a pattern of supporting different backends/accelerators in a more modular way instead of entangling everything together. The downside is that extensions are not complete and there are things that cannot be done at the moment. One example is defining new types in the main module (without `evaling`). This is manifest in this PR, where I pushed the definition of `MPISendRecvGraphContext` and `MPIPersistentSendRecvGraphContext` to the extension. A second downside is that extensions are a little brittle right now. For example, there can be errors in precompilation. That said, supporting backends like CUDA is precisely what extensioins were designed for.

Closes #74 
Closes #4 
Closes #69

### CODE REVIEW

**Review by charleskawczynski — APPROVED** [review]
Looks great!

**Comment by Sbozzolo** [comment]
I added a second commit to try to output more informative log/error messages. @charleskawczynski could you have a second look?

I also tried testing this in ClimaCore/ClimaAtmos, but it is a big pain. ClimaComms has tendrils everywhere, and bumping to 0.6 makes updating really tricky because of circular dependencies of the test environments.

To run with ClimaAtmos, I had to check out all the branches and dev this way:
```
(examples) pkg> dev ../ClimaComms.jl/ ../ClimaUtilities.jl/ ../RRTMGP.jl/ ../ClimaCore.jl/ ../ClimaTimeSteppers.jl/
```
I also had to comment out `ClimaCoreTempestRemap`.

I couldn't get this to work on buildkite because I need to dev everything at the same time and I couldn't make that work.

I tried with ClimaAtmos on 2 gpus on the cluster with:
```
CLIMACOMMS_DEVICE="CUDA" CLIMACOMMS_CONTEXT="MPI" $MPITRAMPOLINE_MPIEXEC -n 2 julia --project=examples examples/hybrid/driver.jl --config_file config/mpi_configs/mpi_make_restart.yml
```
And this works

I couldn't perform a more comprehesive test, but things are looking code from the interactive tests I've performed. So, I think we should release ClimaComms 0.6 and move the various other packages one by one. For ClimaUtilities, ClimaTimesteppers, RRTMGP, ClimaCoreTempestRemap the update is easy and non breaking, so we should move them first. Then, we can probably move to ClimaCore and release 0.14, and re-update ClimaUtilities, ClimaTimeSteppers, RRTMGP, ClimaCoreTempestRemap to support ClimaCore 0.14. Finally, we can update ClimaCoupler, ClimaAtmos, and ClimaLand to ClimaCore 0.14 and ClimaComms 0.6

File: src/loading.jl [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    return quote
+        @static if $mpi_is_required() && $mpi_ext_is_not_loaded()
+            try
+                import MPI
+                @info "Loaded MPI.jl"
+            catch
+                error(
+                    "Cannot load MPI.jl. Make sure it is included in your environment stack.",
+                )
+            end
```
Comment (charleskawczynski) [inline src/loading.jl:37]:
```suggestion
            import MPI
            @info "Imported MPI"
```
Why do we need a try-catch here?

Comment (Sbozzolo) [reply src/loading.jl]: Removed!

**Pull request merged**